src/browser/connection/index.js

```diff
@@ -0,0 +1,198 @@
+        });
+
+        this.browserInfo.provider.registerEvaluator(this.id, code => new Promise(resolve => {
+            this.providerEvalQueue.push({ code, resolve });
+        }));
+
+        this.browserConnectionGateway.startServingConnection(this);
+
+        this._runBrowser();
+    }
+
+    static _generateId () {
+        return shortId.generate();
+    }
+
+    _runBrowser () {
+        // NOTE: Give caller time to assign event listeners
+        process.nextTick(async () => {
+            try {
+                await this.browserInfo.provider.openBrowser(this.id, this.browserInfo.descriptor, this.url);
+                this.opened = true;
+            }
+            catch (err) {
+                this.emit('error', err);
+            }
+        });
+    }
+
+    async _closeBrowser () {
+        if (!this.idle)
+            await promisifyEvent(this, 'idle');
+
+        await this.browserInfo.provider.closeBrowser(this.id, { title: this.idleUrl, url: this.idleUrl });
+    }
+
+    _forceIdle () {
+        if (!this.idle) {
+            this.switchingToIdle = false;
+            this.idle            = true;
+            this.emit('idle');
+        }
+    }
+
+    _waitForHeartbeat () {
+        this.heartbeatTimeout = setTimeout(() => {
+            this.emit('error', new GeneralError(MESSAGE.browserDisconnected, this.userAgent));
+        }, this.HEARTBEAT_TIMEOUT);
+    }
+
+    _popNextTestRunUrl () {
+        while (this.hasQueuedJobs && !this.currentJob.hasQueuedTestRuns)
+            this.jobQueue.shift();
+
+        return this.hasQueuedJobs ? this.currentJob.popNextTestRunUrl() : null;
+    }
+
+    get hasQueuedJobs () {
+        return !!this.jobQueue.length;
+    }
+
+    get currentJob () {
+        return this.jobQueue[0];
+    }
+
+    // API
+    addJob (job) {
+        this.jobQueue.push(job);
+    }
+
+    removeJob (job) {
+        remove(this.jobQueue, job);
+    }
+
+    close () {
+        if (this.closed)
+            return;
+
+        clearTimeout(this.heartbeatTimeout);
+
+        this._closeBrowser()
+            .then(() => {
+                this.browserConnectionGateway.stopServingConnection(this);
+
+                this.ready  = false;
+                this.closed = true;
+
+                this.emit('closed');
+            });
+    }
+
+    establish (userAgent) {
+        this.ready     = true;
+        this.userAgent = parseUserAgent(userAgent).toString();
+
+        this._waitForHeartbeat();
+        this.emit('ready');
+    }
+
+    heartbeat () {
+        clearTimeout(this.heartbeatTimeout);
+        this._waitForHeartbeat();
+    }
+
+    renderIdlePage () {
+        return Mustache.render(IDLE_PAGE_TEMPLATE, {
+            userAgent:    this.userAgent,
+            statusUrl:    this.statusUrl,
+            heartbeatUrl: this.heartbeatUrl,
+            evalUrl:      this.evalUrl
+        });
+    }
+
+    getEvalStatus () {
+        var promisedEval = this.providerEvalQueue[0];
```

Comment:
OK, I thought so. Can we have better naming for this then? Something like `runBrowserInitializationScript` or smthg like that?